pkg/gardenadm/botanist/ssh.go

```diff
@@ -59,20 +60,33 @@ func (b *AutonomousBotanist) ConnectToMachine(ctx context.Context, index int) (*
 }
 
 func (b *AutonomousBotanist) sshAddressForMachine(machine *machinev1alpha1.Machine) string {
-	// For now, we expect that the Bastion can connect to the control plane machine via its hostname and that the hostname
-	// is equal to the machine's name.
-	// More cases will be supported once https://github.com/gardener/machine-controller-manager/pull/1012 gets released.
-	// TODO(timebertt): prefer Machine.status.addresses if present once mcm has been updated in gardener
-	machineHostname := machine.Name
-
-	// Until machine-controller-manager-provider-local reports the correct hostname/IP in Machine.status.addresses, we
-	// prefix the machine name with "machine-" because we know that this is the hostname of local machines.
-	// TODO(timebertt): drop this shortcut once https://github.com/gardener/gardener/pull/12489 has been merged.
-	if b.Shoot.GetInfo().Spec.Provider.Type == "local" {
-		machineHostname = "machine-" + machineHostname
+	var hostname, ip string
+
+	for _, address := range machine.Status.Addresses {
+		switch address.Type {
+		case corev1.NodeInternalIP:
+			ip = address.Address
+		case corev1.NodeInternalDNS:
+			hostname = address.Address
+
+		// internal names have priority, as we jump via a bastion host
+		case corev1.NodeExternalIP:
+			if ip == "" {
+				ip = address.Address
+			}
+		case corev1.NodeExternalDNS:
```

Comment:
What about `Hostname`? We should probably respect it, given that this is the only address type that is always present.